### PR TITLE
engine: Port MVCCScanner to Go, implement MVCCScan for pebble

### PR DESCRIPTION
**Pull request by itsbilal**
This change ports the MVCC Scanner from libroach to go, for use
with Pebble's iterators. One half of the work involved in #39674.
Also updates pebbleIterator to call it in pebbleIterator.MVCC*.

First commit is #41199 , which should merge soon.

Release note: None

### CODE REVIEW

**Comment by itsbilal** [comment]
Dependency graph of related PRs:
```
#41199 ---------------------  #41452  --------------------------- #41455
(pebble engine)       \      (pebble MVCC scanner)              (MVCC test refactor)
                       \        (this PR)
                        \
                         ----- #41453
                            (CLI changes)
```

**Comment by petermattis** [comment]
Thanks for splitting up #41199. I'm going to hold off on reviewing this until the other PRs are in.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
haven't finished reading, but sending some initial questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-:-LqvNi_I358bfZqgwFJ3:b-7aghhc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/pebble_iterator.go, line 382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvLX4q3b_G1By4Q24-:-LqvLX4q3b_G1By4Q240:b-5vczd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_iterator.go#L382)):*
> ```Go
> func (p *pebbleIterator) MVCCScan(
> 	start, end roachpb.Key, max int64, timestamp hlc.Timestamp, opts MVCCScanOptions,
> ) (kvData []byte, numKVs int64, resumeSpan *roachpb.Span, intents []roachpb.Intent, err error) {
> ```

unrelated question: why does an Intent contain a Span with an EndKey and not just the Key? That is, when are intents about a range of keys?
___
*[pkg/storage/engine/pebble_iterator.go, line 437 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvIdmm26cql1FsA_e5:-LqvIdmm26cql1FsA_e6:b4lrba9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_iterator.go#L437)):*
> ```Go
> 				EndKey: p.mvccScanner.curKey,
> 			}
> 			resumeSpan.EndKey = resumeSpan.EndKey.Next()
> ```

is this since p.mvccScanner.curKey is not included in kvData and so we need to see it when resuming?
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvK1YT3thB3-DIHFL3:-LqvK1YT3thB3-DIHFL4:b-2v1esv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L50)):*
> ```Go
> 	} else {
> 		oldRepr := p.repr
> 		p.repr = make([]byte, len(oldRepr)+lenToAdd)
> ```

don't we want to allocate something bigger than exactly what we need to accommodate this put()?
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 79 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvMbtD8EHUCqXin7Pu:-LqvMbtD8EHUCqXin7Pv:b97uadg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L79)):*
> ```Go
> 	meta                        enginepb.MVCCMetadata
> 	inconsistent, tombstones    bool
> 	ignoreSeq, checkUncertainty bool
> ```

can you add a reference comment here for the bools coming from MVCC{Scan,Get}Options, and for the others what they mean?
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 87 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvNR01CGY-6w2PLUBB:-LqvNR01CGY-6w2PLUBC:bg91ol8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L87)):*
> ```Go
> 	curTS                       hlc.Timestamp
> 	results                     pebbleResults
> 	intents                     pebble.Batch
> ```

are we putting all the intents we find in this batch?

**Review by petermattis — COMMENTED** [review]
@itsbilal Did an initial pass. Much of `pebbleMVCCScanner` looks good, though I discovered that we should change the API for `Iterator.MVCC{Scan,Get}`. I don't think that API change will be particularly difficult and it will remove an unnecessary bit of encoding and decoding from the Pebble scan code path. Let me know if it isn't clear what I'm asking for.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-:-Lr0ReJU59r3UTdHuoqr:b-u92efo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/pebble_iterator.go, line 382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvLX4q3b_G1By4Q24-:-Lr0QVVCB_ekPgJQNKa9:b-f5voau) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_iterator.go#L382)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

unrelated question: why does an Intent contain a Span with an EndKey and not just the Key? That is, when are intents about a range of keys?
</blockquote></details>

Intents are placed on specific records, but the `roachpb.Intent` proto is also used to represent key spans that contain intents. My understanding is that this is used in our handling of large transactions. Rather than precisely track the list of records touched in the transaction, at some point we switch over to tracking spans of keys that contain intents. Nathan or Tobi would likely know more of the details.
___
*[pkg/storage/engine/pebble_iterator.go, line 382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0QtIH5R18-lxZBedW:-Lr0QtIH5R18-lxZBedX:bohfumb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_iterator.go#L382)):*
> ```Go
> func (p *pebbleIterator) MVCCScan(
> 	start, end roachpb.Key, max int64, timestamp hlc.Timestamp, opts MVCCScanOptions,
> ) (kvData []byte, numKVs int64, resumeSpan *roachpb.Span, intents []roachpb.Intent, err error) {
> ```

Per my other comment, I think we should change API of `Iterator.MVCCScan` to have the same return values as `engine.MVCCScan`. Then we can have the Pebble version directly scan into `[]roachpb.KeyValue` and `[]roachpb.Intent` avoiding an unnecessary copy. This API change should be done in a separate precursor PR so we can review it in isolation.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvK1YT3thB3-DIHFL3:-Lr0K9WgC8UITHrF0nP3:b-etz0bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L50)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

don't we want to allocate something bigger than exactly what we need to accommodate this put()?
</blockquote></details>

Yes, though I suspect we don't need `pebbleResults` at all. The reason for its existence when using RocksDB is to efficiently pass back the results across the Cgo boundary. With Pebble we don't have that concern. It will require a bit of restructuring in order to do this. `MVCCScan` right now expects data that can be passed to `MVCCScanDecodeKeyValue`. We'll have to refactor that code to hide the existing of `MVCCScanDecodeValue` in the RocksDB engine implementation.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0M0qw5nkQYztOhaAy:-Lr0M0qw5nkQYztOhaAz:b-n9p4hy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L111)):*
> ```Go
> 	mvccKey := MVCCKey{key, hlc.Timestamp{}}
> 	p.keyBuf = EncodeKeyToBuf(p.keyBuf[:0], mvccKey)
> 	p.parent.SeekGE(p.keyBuf)
> ```

I think this is mirroring the logic from the C++ code, but I suspect this call can be `SeekLT` which would remove the need for the subsequent `SeekLT` call and the call to `.Prev`.

With Pebble, we should use the return value from the call rather than calling `Valid` (when possible):

```
if !p.parent.SeekLT(p.keyBuf) {
  return
}
```
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 243 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0MfpF6N9YTp-m5nUZ:-Lr0MfpF6N9YTp-m5nU_:b-pdoica) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L243)):*
> ```Go
> 	}
> 
> 	if bytes.Compare(mvccKey.Key, p.curKey) == 0 {
> ```

Better/faster to use `bytes.Equal` if all you're checking for is equality.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0Mo4lBWV8ydL4d_Tf:-Lr0Mo4lBWV8ydL4d_Tg:b8ixfiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L256)):*
> ```Go
> 		p.decrementItersBeforeSeek()
> 		return
> 	} else if gotToPrevious {
> ```

Nit: no need for the `else` here as the previous block unconditionally returned.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 276 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0Mvyk91a5Bd0E71vH:-Lr0Mvyk91a5Bd0E71vI:bjpt1jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L276)):*
> ```Go
> 	mvccKey, err := DecodeMVCCKey(p.parent.Key())
> 
> 	for iterCount >= 0 && p.parent.Valid() && err == nil && bytes.Compare(mvccKey.Key , p.curKey) == 0 {
> ```

`s/bytes.Compare/bytes.Equal/g`
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 286 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0MyMB7aLJQnN3kZdC:-Lr0MyMB7aLJQnN3kZdD:bjpt1jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L286)):*
> ```Go
> 	}
> 
> 	if bytes.Compare(mvccKey.Key, p.curKey) == 0 {
> ```

`s/bytes.Compare/bytes.Equal/g`
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 330 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0N27C3I_2oWKVtJf-:-Lr0N27C3I_2oWKVtJf0:bnci2d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L330)):*
> ```Go
> 		p.decrementItersBeforeSeek()
> 		return
> 	} else {
> ```

Nit: no need for the else since the previous block returns unconditionally.

**Review by petermattis — COMMENTED** [review]
> I don't think that API change will be particularly difficult and it will remove an unnecessary bit of encoding and decoding from the Pebble scan code path. Let me know if it isn't clear what I'm asking for.

Well, this turned out to be bad advice. `MVCCScan` returns `[]roachpb.KeyValue`, but the primary KV operation uses `MVCCScanToBytes` now. Looks like we'll need to continue with the use of `pebbleResults`, though one possible tweak I noticed is that `MVCCScanToBytes` returns a single `[]byte` of the KV data, while the `ScanResponse` has a slice of slices. It may be useful to change `MVCCScanToBytes` to return a `[][]byte` so that we can avoid an unnecessary copy.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-:-Lr0fbVU8XefSoOt0Gcq:b-rnbwtr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, and @sumeerbhola)

**Comment by itsbilal** [comment]
Um, I don't recall pressing the close button. Don't know why that happened. Reopening.

**Review by itsbilal — COMMENTED** [review]
Addressed comments, except for the ones around changing the `MVCC{Scan, Get}` APIs due to the existence of `MVCCScanToBytes`. Thanks for reviewing, and another look would be appreciated!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-:-LrAOT-y3eBvrZhrSsag:b-3okqdp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/pebble_iterator.go, line 437 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvIdmm26cql1FsA_e5:-Lr9nXAE3Sk3g1GPlLmf:b95nkfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_iterator.go#L437)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this since p.mvccScanner.curKey is not included in kvData and so we need to see it when resuming?
</blockquote></details>

Yes, that's correct. We haven't added `curKey` to the result set and need to include it going forward. `EndKey` is always treated as exclusive, whether `Reverse` is true or not. I'll add a comment to clarify this.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvK1YT3thB3-DIHFL3:-Lr9oK-L3fXCvMYQcmTR:bwtr59l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L50)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yes, though I suspect we don't need `pebbleResults` at all. The reason for its existence when using RocksDB is to efficiently pass back the results across the Cgo boundary. With Pebble we don't have that concern. It will require a bit of restructuring in order to do this. `MVCCScan` right now expects data that can be passed to `MVCCScanDecodeKeyValue`. We'll have to refactor that code to hide the existing of `MVCCScanDecodeValue` in the RocksDB engine implementation.
</blockquote></details>

This is setting `len`, not `cap`, on the newly created slice. My understanding is Go will create a greater capacity each time this happens. But I might be wrong.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 79 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvMbtD8EHUCqXin7Pu:-Lr9sx5g96_H--y2mBID:b-j5fgsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L79)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you add a reference comment here for the bools coming from MVCC{Scan,Get}Options, and for the others what they mean?
</blockquote></details>

The package-level `MVCCScan` has a comment explaining these in detail already. Added a comment to refer readers there.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 87 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvNR01CGY-6w2PLUBB:-Lr9pSn19UeSPnGMaHPJ:b33zl75) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L87)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

are we putting all the intents we find in this batch?
</blockquote></details>

Yes, any intents we encounter from other transactions in `seekVersion` go in here.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0M0qw5nkQYztOhaAy:-Lr9uDWiE3WhWYhb341o:buk1zi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L111)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think this is mirroring the logic from the C++ code, but I suspect this call can be `SeekLT` which would remove the need for the subsequent `SeekLT` call and the call to `.Prev`.

With Pebble, we should use the return value from the call rather than calling `Valid` (when possible):

```
if !p.parent.SeekLT(p.keyBuf) {
  return
}
```
</blockquote></details>

Done. Good call - this does simplify this method a lot.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 243 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0MfpF6N9YTp-m5nUZ:-LrAKmxf1Jip3hrKJ9cc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L243)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Better/faster to use `bytes.Equal` if all you're checking for is equality.
</blockquote></details>

Done.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 276 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0Mvyk91a5Bd0E71vH:-LrAL9lH3FLsDoRyJcrw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L276)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/bytes.Compare/bytes.Equal/g`
</blockquote></details>

Done.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 286 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-Lr0MyMB7aLJQnN3kZdC:-LrALNSY6tQCoSCR2_Wp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L286)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/bytes.Compare/bytes.Equal/g`
</blockquote></details>

Done.

**Review by petermattis — COMMENTED** [review]
I need to do another pass on this PR, but it generally looks good.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-:-LrAUSCt-OkhcByfYIPI:bbbu4fj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, and @sumeerbhola)

___
*[pkg/storage/engine/pebble_iterator.go, line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LrATUkV4YawUqtICQ6T:-LrATUkV4YawUqtICQ6U:b-84i7k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/62ff40792b35860751a55e59cf1964e6ee1ccd9d/pkg/storage/engine/pebble_iterator.go#L43)):*
> ```Go
> 	prefix bool
> 	// MVCC Scanner struct to store variables for one call to MVCCGet / MVCCScan.
> 	mvccScanner pebbleMVCCScanner
> ```

I wonder if this should be pool allocated instead of always embedded within `pebbleIterator`.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvK1YT3thB3-DIHFL3:-LrATcS_4rKG27_iZYCG:bhebfuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L50)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

This is setting `len`, not `cap`, on the newly created slice. My understanding is Go will create a greater capacity each time this happens. But I might be wrong.
</blockquote></details>

Go grows the backing store when needed when you use `append`, but it can't do that with a call to `make` as it doesn't know the size of the old backing store. It is difficult to use `append` in this instance, though because we don't have the new data to append. What is usually done is to perform the exponential growth ourselves using something like the code in `util/bufalloc`. Note that you can't use `bufalloc` directly here because that doesn't create a contiguous buffer (which is what you want). On the other hand, I think we could change the API for `MVCCScan` so that the return value is a slice of buffers of exponentially increasing size. But let's leave that to a follow-on PR. (Add a TODO in this one).

**Review by itsbilal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/pebble_iterator.go, line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LrATUkV4YawUqtICQ6T:-LrAbpJG7z7VWAHC6PvY:b6t7lgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/62ff40792b35860751a55e59cf1964e6ee1ccd9d/pkg/storage/engine/pebble_iterator.go#L43)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I wonder if this should be pool allocated instead of always embedded within `pebbleIterator`.
</blockquote></details>

The benefits of that would be more evident if we have a couple frequent `pebbleIterator` use-cases that don't call MVCCGet or MVCCScan, right? Is that the case right now? 
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvK1YT3thB3-DIHFL3:-LrAas-qC_7LxFbKS9Ar:bvlsmb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L50)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Go grows the backing store when needed when you use `append`, but it can't do that with a call to `make` as it doesn't know the size of the old backing store. It is difficult to use `append` in this instance, though because we don't have the new data to append. What is usually done is to perform the exponential growth ourselves using something like the code in `util/bufalloc`. Note that you can't use `bufalloc` directly here because that doesn't create a contiguous buffer (which is what you want). On the other hand, I think we could change the API for `MVCCScan` so that the return value is a slice of buffers of exponentially increasing size. But let's leave that to a follow-on PR. (Add a TODO in this one).
</blockquote></details>

Done. Added a TODO on returning non-contiguous buffers here and in `MVCCScan*`.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, and @sumeerbhola)

___
*[pkg/storage/engine/pebble_iterator.go, line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LrATUkV4YawUqtICQ6T:-LrAcil6D2NvkU9edgqe:beohm76) ([raw file](https://github.com/cockroachdb/cockroach/blob/62ff40792b35860751a55e59cf1964e6ee1ccd9d/pkg/storage/engine/pebble_iterator.go#L43)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

The benefits of that would be more evident if we have a couple frequent `pebbleIterator` use-cases that don't call MVCCGet or MVCCScan, right? Is that the case right now? 
</blockquote></details>

I believe most KV level operations do not explicitly use `MVCC{Get,Scan}`. But rather than speculate, you could instrument the code and see. Does a Put operation use `MVCCGet`?
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvK1YT3thB3-DIHFL3:-LrAd7hg3pWYmumALJv4:buf3bus) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L50)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Done. Added a TODO on returning non-contiguous buffers here and in `MVCCScan*`.
</blockquote></details>

You should still do the exponential buffer growth here, even if only a single buffer is being used. The way this code is currently written you'll be growing the buffer on every single call to `pebbleResults.put`.

**Review by itsbilal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/pebble_iterator.go, line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LrATUkV4YawUqtICQ6T:-LrBCl1EBt_zGaVGbtCp:byp14sc) ([raw file](https://github.com/cockroachdb/cockroach/blob/62ff40792b35860751a55e59cf1964e6ee1ccd9d/pkg/storage/engine/pebble_iterator.go#L43)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I believe most KV level operations do not explicitly use `MVCC{Get,Scan}`. But rather than speculate, you could instrument the code and see. Does a Put operation use `MVCCGet`?
</blockquote></details>

Put uses `mvccGetInternal` which follows a Get-oriented go-only path. We should explore using it for `MVCCGet`.

Either way, on my small read/write balanced workload it seemed like most iterators were not seeing an MVCCScanner operation on them, so I made the change to pool allocate scanner structs.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-LqvK1YT3thB3-DIHFL3:-LrBEjSv9rXrpB9MCRjm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f808828b551ce06795ebdd70acac9b4607439de8/pkg/storage/engine/pebble_mvcc_scanner.go#L50)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You should still do the exponential buffer growth here, even if only a single buffer is being used. The way this code is currently written you'll be growing the buffer on every single call to `pebbleResults.put`.
</blockquote></details>

Done.

**Review by petermattis — APPROVED** [review]
:lgtm:

I'm going to do another pass after this is merged comparing the C++ vs Go logic. Will be easier to do that once this is merged.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41452#-:-LrJjl-TB5JHWPZzWKgE:b-dj9e7f)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajkr and @sumeerbhola)

**Pull request merged**
